src/Aquifer.Public.API/Endpoints/Resources/Updates/List/Endpoint.cs

```diff
@@ -16,23 +18,73 @@ public override void Configure()
         {
             s.Summary = "Get resource ids that are new or updated since the given UTC timestamp.";
             s.Description =
-                "For a given UTC timestamp and optional language id, get a list of resource ids that are new or have been updated since the provided timestamp. This is intended for users who are storing Aquifer data locally and want to fetch new content.";
+                "For a given UTC timestamp, get a list of resource ids that are new or have been updated since the provided timestamp. This is intended for users who are storing Aquifer data locally and want to fetch new content.";
         });
     }
 
     public override async Task HandleAsync(Request req, CancellationToken ct)
     {
-        Response.Items = await dbContext.ResourceContentVersions
-            .Where(x => (!req.LanguageId.HasValue || x.ResourceContent.LanguageId == req.LanguageId) &&
-                x.IsPublished &&
-                x.Updated >= req.Timestamp)
+        var query = await GetQuery(req, ct);
+        var totalCount = await GetTotalResourceCountAsync(req, query, ct);
+
+        if (totalCount == 0)
+        {
+            await SendOkAsync(new Response
+            {
+                Offset = req.Offset
+            }, ct);
+            return;
+        }
+
+        var items = await GetResourceUpdatesAsync(req, query, ct);
+        Response = new Response
+        {
+            Items = items,
+            TotalItemCount = totalCount,
+            Offset = req.Offset
+        };
+    }
+
+    private async Task<List<ResponseContent>> GetResourceUpdatesAsync(Request req, IQueryable<ResourceContentVersionEntity> query,
+        CancellationToken ct)
+    {
+        var languageCodeByIdMap = await cachingLanguageService.GetLanguageCodeByIdMapAsync(ct);
+        return await query
+            .OrderByDescending(r => r.Updated)
+            .Skip(req.Offset)
+            .Take(req.Limit)
             .Select(x => new ResponseContent
             {
                 ResourceId = x.ResourceContentId,
                 LanguageId = x.ResourceContent.LanguageId,
+                LanguageCode = languageCodeByIdMap[x.ResourceContent.LanguageId],
                 UpdateType = x.Version == 1 ? ResponseContentUpdateType.New : ResponseContentUpdateType.Updated,
                 Timestamp = x.Updated
             })
             .ToListAsync(ct);
     }
+
+    private async Task<int> GetTotalResourceCountAsync(Request req, IQueryable<ResourceContentVersionEntity> query, CancellationToken ct)
+    {
+        var totalCount = await query.CountAsync(ct);
+        if (req.Offset >= totalCount && req.Offset != 0)
+        {
+            ThrowError($"Offset of {req.Offset} equals or exceeds total of {totalCount}");
+        }
+
+        return totalCount;
+    }
+
+    private async Task<IQueryable<ResourceContentVersionEntity>> GetQuery(Request req, CancellationToken ct)
+    {
+        var (_, _, validLanguageId) =
```

Comment:
This will ignore invalid language IDs like they weren't passed at all.  Do we want to return an error instead?  IMO as a consumer I would expect a bad request rather than data not being what I expected.

See https://github.com/BiblioNexusStudio/aquifer-server/blob/48af2303af99152adc2bbfa21676583a59f814d1/src/Aquifer.Public.API/Endpoints/Resources/Collections/Get/Endpoint.cs#L39-L46 for an example of throwing errors (though that route allows multiple language IDs or Codes so it will be simpler here).

Reply:
Not sure what you mean by "weren't passed at all".  I had thought that we needed this to be open to the possibility of one or both of them being null.

I've tested putting in an invalid language code, and it just returns nothing. In that case it probably would be good for consumer and end user to know that their language code is invalid if they were to pass in something like 'foo', and the validator makes sure that the languageId is between 1 and 32 if it were passed in.

Reply:
> I've tested putting in an invalid language code, and it just returns nothing. In that case it probably would be good for consumer and end user to know that their language code is invalid if they were to pass in something like 'foo', and the validator makes sure that the languageId is between 1 and 32 if it were passed in.

The validator doesn't handle all scenarios.  For example, a `LanguageCode` of `TST` will pass validation but will result in no results being returned here which is unexpected IMO.  You could instead do:
```
 if (isInvalidLanguageCode) // from validation method return value
 { 
     ThrowError(r => r.LanguageCode, $"Invalid 'Language Code': {req.LanguageCode}"); 
 } 
```
Likewise, not all language IDs between 1 and 32 are valid.

Reply:
> Sending language is optional for this endpoint, one or the other isn't required. Is that what you were getting at Nate?

Sorry for the confusion.  I mean the scenario where one of language code or language ID is passed but that code or ID is not in the DB.  In that case the current PR will ignore the invalid ID or code instead of erroring.

Reply:
I agree that it should return a relevant error message in that case.

Reply:
It spits out an error now (just needs a tweak),  so make sure you pull the latest if you're testing.

Reply:
> The validator doesn't handle all scenarios. For example, a LanguageCode of TST will pass validation but will result in no results being returned here which is unexpected IMO. You could instead do:

> if (isInvalidLanguageCode) // from validation method return value
 { 
     ThrowError(r => r.LanguageCode, $"Invalid 'Language Code': {req.LanguageCode}"); 
 } 
 
I can't find `isInvalidLanguageCode`  being returned in the validator, or anywhere in the `CachingLanguageService` file, but I have used `isValidLanguageCode / Id` for the error messages

Reply:
Sorry, I meant `isValid...` not `isInvalid`.

See [my original comment](https://github.com/BiblioNexusStudio/aquifer-server/pull/502#discussion_r1815544627) for getting back an `isValidLanguageCode` and `isValidLanguageId` result (instead of using the discard variable `_`):
```
var (isValidLanguageId, isValidLanguageCode, validLanguageId) = await ValidateLanguageIdOrCodeAsync(req.LanguageId, req.LanguageCode, shouldRequireInput: false, ct);
```